library/ssl_msg.c

```diff
@@ -6058,6 +6058,81 @@ int mbedtls_ssl_write(mbedtls_ssl_context *ssl, const unsigned char *buf, size_t
     return ret;
 }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA) && defined(MBEDTLS_SSL_CLI_C)
+int mbedtls_ssl_write_early_data(mbedtls_ssl_context *ssl,
+                                 const unsigned char *buf, size_t len)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    const struct mbedtls_ssl_config *conf;
+    int written_data_len = 0;
+
+    MBEDTLS_SSL_DEBUG_MSG(2, ("=> write early_data"));
+
+    if (ssl == NULL || (conf = ssl->conf) == NULL) {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+    }
+
+    if ((!mbedtls_ssl_conf_is_tls13_enabled(conf)) ||
+        (conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM) ||
+        (conf->early_data_enabled != MBEDTLS_SSL_EARLY_DATA_ENABLED)) {
+        return MBEDTLS_ERR_SSL_CANNOT_WRITE_EARLY_DATA;
+    }
+
+    if (ssl->tls_version != MBEDTLS_SSL_VERSION_TLS1_3) {
+        return MBEDTLS_ERR_SSL_CANNOT_WRITE_EARLY_DATA;
+    }
+
+    /*
+     * If we are at the beginning of the handshake, advance the handshake just
+     * enough to be able to send early data if possible. That way, we can
+     * guarantee that when starting the handshake with this function we will
+     * send at least one record of early data.
+     * Otherwise, resume the handshake and if the handshake sequence stops
+     * waiting for some message from the server, send early data if we can.
+     */
```

Comment:
> I tend to reason in terms of handshake steps. So, I understand that if we haven't sent our ClientHello yet, clearly we want to send that first, and that's what calling `mbedtls_ssl_handshake_step()` seems to be doing in the first branch. But I don't really understand why we want to call that when status is `SENT` - to me that means we've sent the extension hence the ClientHello already? Or is it that we've written it to the buffer but it hasn't actually been sent out on the network yet (in which case the name would be confusing).

That's the difference between MBEDTLS_SSL_EARLY_DATA_STATUS_SENT and MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_WRITE. If the middlebox compatibility is enabled, after sending the ClientHello and before switching to the early data key for outbound traffic we have to send in clear the dummy CCS and that's one more handshake step.

Reply:
> Regarding the second branch: so, despite its name, this function's goal is not just to sent early data, it's to complete the handshake if possible, and if not because we're waiting for the server, sent early data. Right?

I wouldn't say that to complete the handshake is a goal of this API, it is rather to keep progressing the handshake as long as we are in the window where early data can be sent (status CAN_WRITE and ACCEPTED). We do not want the client to keep sending early data and the handshake to be stalled. Thus each time the API is called in the "write window" we try to advance as much as possible the handshake, basically up to the point where we wait for some server message or we wait for some data to be actually sent.

Reply:
> Note: with the second branch I'm having the same problem mentally mapping `EARLY_DATA_STATUS_VALUES` to handshake states. I'm thinking `CAN_WRITE` right after we've sent the ClientHello and before we received EncryptedExtensions from the server, and `ACCEPTED` is after we've receive the extension from the server. Is that correct?

Yes that's correct but maybe `CAN_WRITE` right after we've sent the ClientHello as we may still have to send a dummy CCS (see first response above).

Reply:
> We do not want the client to keep sending early data and the handshake to be stalled.

Ah, good point. This makes the API more user-friendly in that users can just keep calling this function until it returns an error. Otherwise they'd need to be sure to call `handshake()` in-between calls to this function, which would be a bit awkward, and we're in a better position than users to know when that's useful. Thanks for clarifying that (and answering my other questions).